pkg/serviceaccounts/controllers/mutation_cache.go

```diff
@@ -0,0 +1,88 @@
+package controllers
+
+import (
+	"github.com/golang/glog"
+	lru "github.com/hashicorp/golang-lru"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+)
+
+// MutationCache is able to take the result of update operations and stores them in an LRU
+// that can be used to provide a more current view of a requested object.  It requires interpretting
+// resourceVersions for comparisons
+type MutationCache interface {
+	GetByKey(key string) (interface{}, bool, error)
+	Mutation(interface{})
+}
+
+type ResourceVersionComparator interface {
+	CompareResourceVersion(lhs, rhs runtime.Object) int
+}
+
+// NewEtcdMutationCache gives back a MutationCache that understands how to deal with etcd backed objects
+func NewEtcdMutationCache(backingCache cache.Store) MutationCache {
+	lru, err := lru.New(100)
+	if err != nil {
+		// errors only happen on invalid sizes, this would be programmer error
+		panic(err)
+	}
+
+	return &mutationCache{
+		backingCache:  backingCache,
+		mutationCache: lru,
+		comparator:    etcd.APIObjectVersioner{},
+	}
+}
+
+type mutationCache struct {
+	backingCache  cache.Store
+	mutationCache *lru.Cache
+
+	comparator ResourceVersionComparator
+}
+
+func (c *mutationCache) GetByKey(key string) (interface{}, bool, error) {
+	obj, exists, err := c.backingCache.GetByKey(key)
```

Comment:
> So mutation cache can contain entries not in the backing cache, but as long as you only ever call GetByKey, you can't observe anything not in the backing cache.

See comment a few lines down.  You can't distinguish between "didn't observe create" and "item was deleted after creation".  I think we should err in the "ignore cache" direction, since erring the other way means you may never see a delete you should.